### DISCUSSION TITLE
Pi Flows: Bringing PowerFx to Raspberry Pi

### DISCUSSION THREAD

**Discussion by eswarpr** [microsoft/Power-Fx#1117] [Show and tell]
Good day all,

I have been working on this quietly in the background and have managed to get it to a state where PowerFx can be used to program for Raspberry Pi. I have made a video explaining the features (as of now) at https://youtu.be/GjT3K5r2H8Y

Just thought I will shout about it a bit since I quite like what I have built with this library. 😀

**Comment by MikeStall** [comment]
This looks great! Please keep us posted on progress.

**Reply by eswarpr** [reply]
Thanks @MikeStall. I have just moved the app forward a bit more with HTTP integration support. But I am not quite happy with what I have done to achieve it. It is "not clean" like I want it to be. Here is the issue:

When I need to allow the app to prompt the user for a symbol (which is available within the context) with a structure that is pretty loose - like we know "Request" as a symbol will be there, but we do not know the shape of "Request.Parameters", his do I do it? How do I go about expressing that? It will be resolved at runtime. I have for now used "DeferredType" and even as per the documentation against that type there is a warning to not use it. What would be the preferred and clean way to achieve something like this?

**Reply by MikeStall** [reply]
Power Fx prefers strong typing - this provides an ideal authoring experience for intellisense completion and errors. For example, if somebody attempts to access a property that doesn't exist - or they just have a typo - strong typing will let us give an immediate error and suggest a correction.

Do you know the schema you're connecting to? 

We do have connector support for importing swagger files. So you can easily connect to a swagger endpoint and get strong typing and a good experience.
https://github.com/microsoft/Power-Fx/tree/main/src/libraries/Microsoft.PowerFx.Connectors
Example tests: https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs 

But if you want to handle an arbitrary Request with unknown properties, you will need to use our "untyped object" support. This is how we handle ParseJson.
See example tests:  https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Json.Tests/ParseJSONTests.cs
The untyped object interfaces are still in preview and we're sorting them out.

**Reply by eswarpr** [reply]
That's great! UntypedValue worked. I will be starting my work on integrating connectors later, but for now, I am trying to integrate it the other way. As in, someone sending a message to the app from Azure. The trouble is they could design the model of the incoming message whichever way they want, and then handle it within the app using a formula fitting that model. I then generate a local model and pass it in as context during the evaluation time. So, basically, we cannot know this before the evaluation time. 

Agree, the intellisense experience is not great, but I have a few ideas spinning around in my head to deal with the schema, but that comes later. 

Thanks for your help Mike. I love working with this library.

**Reply by MikeStall** [reply]
For untyped object, feel free to leave feedback on #1140.

**Comment by MikeStall** [comment]
@eswarpr -  I saw you presented at: 
[May 25 | Demo: Pi Flows: PowerFx for Internet of T... - Power Platform Community (microsoft.com)](https://powerusers.microsoft.com/t5/Community-Calls-Conversations/May-25-Demo-Pi-Flows-PowerFx-for-Internet-of-Things/m-p/2166147#M87) . Congrats! Any updates to share here?

**Reply by eswarpr** [reply]
@MikeStall - From a PowerFx point of view I am just beginning to expand the capability. In the first step, I am going to add support for "Schemas" to allow users to define incoming HTTP request shape for strong typing HTTP processing and adding support for Power Platform Connectors (Custom as well). 

I did a "huge shout out to you and your team" for your help in building the app. Once again, your assistance is much appreciated. Will keep you posted on my progress here.
